File: leetcode-js/103-binary-tree-zigzag-level-order-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    if (!root) {
        return [];
    }

    const result = [];
    let stack = [root];
    let fromLeft = true;

    while (stack.length) {
        const currentStack = stack.slice();
        stack = [];
        const currentRow = [];

        for (let i = currentStack.length - 1; i >= 0; i -= 1) {
            const node = currentStack[i];
            currentRow.push(node.val);
            if (fromLeft) {
                if (node.left) {
                    stack.push(node.left);
                }
                if (node.right) {
                    stack.push(node.right);
                }
            } else {
                if (node.right) {
                    stack.push(node.right);
                }
                if (node.left) {
                    stack.push(node.left);
                }
            }
        }

        result.push(currentRow);
        fromLeft = !fromLeft;
    }

    return result;
};

File: leetcode-java/103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        helper(res, root, 0);
        return res;
    }

    void helper(List<List<Integer>> res, TreeNode node, int level) {
        if (node == null) {
            return;
        }

        if (res.size() == level) {
            res.add(new ArrayList<Integer>());
        }

        if (level % 2 == 0) {
            res.get(level).add(node.val);
        } else {
            res.get(level).add(0, node.val);
        }

        helper(res, node.left, level + 1);
        helper(res, node.right, level + 1);
    }
}
